File: prep/implementation/practice_iterations/merge_intervals_1/MergeIntervals.java
import java.util.*;
import java.util.concurrent.*;
import java.util.function.*;

// O(nlogn) / O(n)
public class MergeIntervals {
  static class Interval {
    int start;
    int end;

    Interval(int start, int end) {
      this.start = start;
      this.end = end;
    }
  }

  public static void main(String[] args) {
    try (Scanner in = new Scanner(System.in)) {
      int tt = in.nextInt();

      while (tt-- > 0) {
        int n = in.nextInt();

        List<Interval> a = new ArrayList<>();
        for (int i = 0; i < n; i++) {
          int s = in.nextInt();
          int e = in.nextInt();

          a.add(new Interval(s, e));
        }

        a.sort((x, y) -> Integer.compare(x.start, y.start));

        int start = a.get(0).start;
        int end = a.get(0).end;
        List<Interval> res = new ArrayList<>();

        for (int i = 1; i < n; i++) {
          if (a.get(i).start <= end) {
            end = Math.max(end, a.get(i).end);
          } else {
            res.add(new Interval(start, end));
            start = a.get(i).start;
            end = a.get(i).end;
          }
        }

        res.add(new Interval(start, end));

        for (var inter : res) {
          System.out.printf("%d %d\n", inter.start, inter.end);
        }
      }
    }
  }
}

File: lc/coreprep/implementation/patterns_reference_implementation/merge_intervals/merge_intervals.cpp
#include <algorithm>
#include <iostream>
#include <vector>

using namespace std;

struct Interval {
  int start;
  int end;

  Interval(int start, int end) : start(start), end(end) {}
};

vector<Interval> merge(vector<Interval> &a) {
  if (a.size() < 2) {
    return a;
  }

  sort(a.begin(), a.end(), [](auto &a, auto &b) { return a.start <= b.start; });

  int start = a[0].start, end = a[0].end;
  vector<Interval> res;

  for (int i = 1; i < a.size(); i++) {
    if (a[i].start <= end) {
      end = max(end, a[i].end);
    } else {
      res.emplace_back(Interval(start, end));
      start = a[i].start;
      end = a[i].end;
    }
  }

  res.emplace_back(Interval(start, end));

  return res;
}

int main() {
  ios::sync_with_stdio(0);
  cin.tie(0);

  int tt, n;
  cin >> tt;

  while (tt--) {
    cin >> n;

    vector<Interval> a;
    int s, e;
    for (int i = 0; i < n; i++) {
      cin >> s >> e;
      a.emplace_back(Interval(s, e));
    }

    auto res = merge(a);
    for (auto inter : res) {
      cout << inter.start << " " << inter.end << "\n";
    }
  }

  return 0;
}